File: jupyter2/problems/tail_recursion/continued_fraction/ml/continued_fraction.ml
(** begin my answer *)

(** end my answer *)


let main () =
  assert (Float.abs (continued_fraction_tail 2.0 100 -. 0.4142136) < 1.0e-6);
  assert (Float.abs (continued_fraction_tail 3.0 100 -. 0.3027756) < 1.0e-6);
  assert (Float.abs (continued_fraction_tail 4.0 100 -. 0.2360680) < 1.0e-6);
  Printf.printf "OK\n"
;;

main()

File: jupyter2/ps/tail_recursion/continued_fraction/go/continued_fraction.go
package main
import "math"
import "fmt"

/** begin my answer */

/** begin hidden */
func continuedFractionTailRec(a float64, m int, xm float64, n int) float64 {
	if m == n  {
		return xm
	} else {
		xm_1 := 1.0 / (a + xm)
		return continuedFractionTailRec(a, m + 1, xm_1, n)
	}
}

func continuedFractionTail(a float64, n int) float64 {
	return continuedFractionTailRec(a, 0, 1.0, n)
}
/** end hidden */
/** end my answer */

func main() {
	if !(math.Abs(continuedFractionTail(2.0, 100) - 0.4142136) < 1.0e-6) { panic("wrong") }
	if !(math.Abs(continuedFractionTail(3.0, 100) - 0.3027756) < 1.0e-6) { panic("wrong") }
	if !(math.Abs(continuedFractionTail(4.0, 100) - 0.2360680) < 1.0e-6) { panic("wrong") }
	fmt.Println("OK")
}
